File: 백준/Gold/15961. 회전 초밥/회전 초밥.py
from collections import defaultdict
n,d,k,c = map(int,input().split())
data = []
for _ in range(n):
    data.append(int(input()))

dic = defaultdict(int)
start = 0
end = k-1

for i in range(k-1):
    dic[data[i]] += 1
# 첫번째 윈도우 세팅하고 시작

res = 0
# 원형이므로 %n 처리 해줘야함 !!! -> 아니면 기존 배열 data 에 이어 붙여서 처리해야함
while end < n + k -1:
    dic[data[end%n]] += 1  # 끝점 추가 후 계산 진행
    if c in dic.keys():
        res = max(res, len(dic))
    else:
        res = max(res, len(dic) + 1)
    # 체크했으니
    dic[data[start%n]] -= 1
    if dic[data[start%n]] == 0:
        dic.pop(data[start%n])
    start += 1
    end += 1  # 끝점 역시 while 문이기에 증가시켜줌

print(res)

File: 백준/Gold/15961. 회전 초밥/회전 초밥.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int n, d, k, c; // ! 접시 수, 초밥 가짓 수, 연속해서 먹는 접시 수, 쿠폰 번호
    static int[] data;
    static int res;
    static Map<Integer, Integer> map = new HashMap<>();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        d = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());

        data = new int[n];
        for (int i = 0; i < n; i++) {
            data[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 0; i < k - 1; i++) {
            map.put(data[i%n], map.getOrDefault(data[i%n], 0) + 1);
        }
        int start = 0;
        int max = 0;
        for (int end = k - 1; end < n + k -1; end++) {
            // 먼저 끝점부터 넣어주기
            map.put(data[end%n], map.getOrDefault(data[end%n], 0) + 1);
            if (!map.keySet().contains(c)) {
                max = Math.max(max, map.size() + 1);
            } else {
                max = Math.max(max, map.size());
            }
            // 시작점 빼기
            map.put(data[start%n], map.get(data[start%n]) -1);
            if (map.get(data[start%n]) == 0) {
                map.remove(data[start%n]);  // 해당 윈도우 제거
            }
            start += 1;
        }

        System.out.println(max);
    }
}